File: paper/validation/run_xafs_ic_params_check.py
# run_xafs_ic_params_check.py -- server-side half of the XAFS IC toggle
# contract check (stage 3 of TASK_XANES_IC_SIM). Consumes the params.ic JSON
# produced by run_xafs_ic_params_check.js (the EXACT object the browser
# assembles) and drives XAFSEngine.run() with it through a stub websocket.
# Asserts the measurement chain ran with the browser-shaped config and the
# result message stays additive-only.
#
# Run: python -X utf8 paper/validation/run_xafs_ic_params_check.py
import asyncio
import json
import math
import os
import sys

ROOT = os.path.join(os.path.dirname(__file__), "..", "..")
sys.path.insert(0, os.path.join(ROOT, "server"))

from sim_engines.xafs_engine import XAFSEngine  # noqa: E402

IC_JSON = os.path.join(os.path.dirname(__file__), "data",
                       "xafs_ic_params_sample.json")


class StubWS:
    def __init__(self):
        self.msgs = []

    async def send(self, s):
        self.msgs.append(json.loads(s))


def run(ic):
    params = {"formula": "Cu", "absorber": "Cu", "edge": "K",
              "eStart": -50, "eEnd": 300, "eStep": 2.0,
              "ppm": 100000, "sampleType": "powder"}
    if ic is not None:
        params["ic"] = ic
    beamline = {"energy_keV": 8.979, "flux": 6.3e12,
                "spot_h_nm": 63.0, "spot_v_nm": 46.0}
    ws = StubWS()
    asyncio.new_event_loop().run_until_complete(
        XAFSEngine().run(ws, params, beamline))
    return [m for m in ws.msgs if m.get("type") == "expt_result"][0]


def main():
    with open(IC_JSON) as f:
        ic = json.load(f)
    print("params.ic from JS:", json.dumps(ic))

    res_on = run(ic)
    res_off = run(None)

    fails = []

    def check(name, cond, detail=""):
        print("  {}  {}  {}".format("PASS" if cond else "FAIL", name, detail))
        if not cond:
            fails.append(name)

    check("ic block present", "ic" in res_on and res_on["ic"]["enabled"] is True)
    check("off has no ic block", "ic" not in res_off
          and "ic_enabled" not in res_off["info"])
    check("info.ic_enabled", res_on["info"].get("ic_enabled") is True)
    m = res_on["ic"]
    check("i0 config echoed", m["i0"]["gas"] == ic["i0"]["gas"]
          and m["i0"]["length_cm"] == ic["i0"]["length_cm"],
          json.dumps(m["i0"]))
    check("i1 config echoed", m["i1"]["gas"] == ic["i1"]["gas"]
          and m["i1"]["length_cm"] == ic["i1"]["length_cm"],
          json.dumps(m["i1"]))
    check("ratio_prefocus applied",
          abs(m["ratio_prefocus"] - ic["ratio_prefocus"]) < 1e-12,
          str(m["ratio_prefocus"]))
    check("dwell applied", m["dwell_s"] == ic["dwell_s"], str(m["dwell_s"]))
    check("flux from beamline SSOT", m["flux_in"] == 6.3e12, str(m["flux_in"]))
    # ratio_prefocus 12.34 -> I0 current must exceed the ratio=1 smoke values
    check("I0 physical (>0, pre-focus boosted)",
          m["i0_A_range"][0] > 0 and m["i0_A_range"][1] > 1e-5,
          "I0 max {:.3e} A".format(m["i0_A_range"][1]))
    ys = [p["y"] for p in res_on["data"]]
    check("y finite", all(map(math.isfinite, ys)))
    check("edge structure", (max(ys) - min(ys)) > 0.5,
          "span {:.3f}".format(max(ys) - min(ys)))
    check("points stay {x,y}",
          {k for p in res_on["data"] for k in p} == {"x", "y"})

    print("RESULT:", "FAIL " + str(fails) if fails else "ALL PASS")
    sys.exit(1 if fails else 0)


if __name__ == "__main__":
    main()

File: paper/validation/run_xafs_ic_params_check.js
// run_xafs_ic_params_check.js -- headless contract check for the XAFS IC
// measurement toggle (stage 3 of TASK_XANES_IC_SIM): loads the experiment
// runner with browser stubs, assembles params.ic via _buildXafsICParams, and
// writes the JSON for the matching server-side check
// (run_xafs_ic_params_check.py) to consume. ES5, node only.
'use strict';
var fs = require('fs');
var path = require('path');
var vm = require('vm');

var ROOT = path.join(__dirname, '..', '..');

// -- Browser stubs (same minimal pattern as run_layout_export_check.js) --
var sandbox = {
  console: console,
  window: {},
  document: {
    getElementById: function () { return null; },
    createElement: function () { return { style: {}, setAttribute: function () {} }; },
    addEventListener: function () {}
  },
  location: { hostname: 'localhost', port: '' },
  WebSocket: function () { this.readyState = 3; },
  setTimeout: function () { return 0; },
  clearTimeout: function () {},
  // state mirrors the IC1 popup + detector IC tab settings under test
  state: {
    energy: 8.979,
    ic1Gas: 'N2', ic1LenCm: 10, ic1PressAtm: 1.0,
    ic1AirBeforeCm: 5, ic1AirAfterCm: 2,
    det_icGas: 'Ar', det_icLenCm: 15
  },
  icLiveChain: function () { return { ratioPreFocus: 12.34 }; }
};
sandbox.globalThis = sandbox;
vm.createContext(sandbox);

var src = fs.readFileSync(
  path.join(ROOT, 'js', 'experiment', '07_experiment_run.js'), 'utf8');
vm.runInContext(src, sandbox, { filename: '07_experiment_run.js' });

var fails = [];
function check(name, cond, detail) {
  console.log('  ' + (cond ? 'PASS' : 'FAIL') + '  ' + name +
    (detail ? '  ' + detail : ''));
  if (!cond) fails.push(name);
}

var build = vm.runInContext('typeof _buildXafsICParams', sandbox);
check('builder defined', build === 'function', build);

var ic = vm.runInContext('_buildXafsICParams(0.5)', sandbox);
check('enabled', ic.enabled === true);
check('dwell', ic.dwell_s === 0.5, String(ic.dwell_s));
check('i0 from IC1 popup state',
  ic.i0 && ic.i0.gas === 'N2' && ic.i0.length_cm === 10 &&
  ic.i0.pressure_atm === 1.0 && ic.i0.air_before_cm === 5 &&
  ic.i0.air_after_cm === 2, JSON.stringify(ic.i0));
check('i1 from det IC tab state',
  ic.i1 && ic.i1.gas === 'Ar' && ic.i1.length_cm === 15 &&
  ic.i1.air_path_cm === 0, JSON.stringify(ic.i1));
check('ratio_prefocus from icLiveChain',
  Math.abs(ic.ratio_prefocus - 12.34) < 1e-12, String(ic.ratio_prefocus));

// dwell fallback + missing-state robustness
var icDef = vm.runInContext('_buildXafsICParams(0)', sandbox);
check('dwell fallback 1.0', icDef.dwell_s === 1.0, String(icDef.dwell_s));
vm.runInContext('state = undefined; icLiveChain = undefined;', sandbox);
var icBare = vm.runInContext('_buildXafsICParams(1.0)', sandbox);
check('robust without state/icLiveChain',
  icBare.enabled === true && icBare.dwell_s === 1.0,
  JSON.stringify(icBare));

// JSON round-trip artifact for the python-side contract check
var out = path.join(__dirname, 'data', 'xafs_ic_params_sample.json');
fs.writeFileSync(out, JSON.stringify(ic, null, 1));
console.log('saved: ' + out);

console.log(fails.length ? 'RESULT: FAIL ' + JSON.stringify(fails)
                         : 'RESULT: ALL PASS');
process.exit(fails.length ? 1 : 0);
